cpp/src/arrow/type_traits.h

```diff
@@ -441,6 +446,16 @@ struct is_8bit_int {
       (std::is_same<UInt8Type, T>::value || std::is_same<Int8Type, T>::value);
 };
 
+template <typename T, typename CType = void>
```

Comment:
Should put this in json_internal.cc unless we anticipate using it elsewhere

Reply:
I do think `IsSignedInt` etc should be removed in favor of this trait or one like it (though it should probably be renamed).

`IsSignedInt` takes either an array or a type as a type argument, which is surprisingly atypical for traits. Furthermore whereas `is_signed_integer` returns false for date and other types which are represented by but not identical to integers `IsSignedInt` returns true by checking only the `c_type`, which leads to `static_assert(IsSignedInt<HalfFloatType>::value, "")`. Finally the declaration of IsSignedInt is far from readable due to nested macro usage

Reply:
https://issues.apache.org/jira/browse/ARROW-6515